### PR TITLE
refactor: `SummarizationAccuracyMetrics` transform to handle multiple target outputs more efficiently

### PR DESCRIPTION
**Pull request by kirupang-code**
- Edited the logic in `SummarizationAccuracyMetrics` so that the `target_output` field wouldn't have to be duplicated in order to be evaluated for each `model_output`. If we have multiple `target_outputs` we take the max score from `self.compute_metric`. 
- Added an additional `target_output_keys_provider` parameter to support a way to compute scores given a key to a list of possible target outputs. 
- Updated the ordering of subclass transforms in order to keep everything consistent with their parent transforms, otherwise parameters would get mixed up. I had to add a default value for `BertScore` because it follows a parameter with a default value.  `input_keys + model_output_keys,
TypeError: can only concatenate str (not "list") to str`

### CODE REVIEW

File: src/fmeval/transforms/summarization_accuracy_metrics.py [hunk 1]
```diff
@@ -126,26 +146,31 @@ class MeteorScore(SummarizationAccuracyMetric):
 
     def __init__(
         self,
-        target_output_keys: List[str],
-        model_output_keys: List[str],
         output_keys: List[str],
-        allow_duplicate_input_keys,
+        model_output_keys: List[str],
+        allow_duplicate_input_keys: bool,
```
Comment (xiaoyi-cheng) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:151]: What's the default for `allow_duplicate_input_keys`? Will this change break existing code?

File: src/fmeval/transforms/summarization_accuracy_metrics.py [hunk 2]
```diff
@@ -36,67 +37,86 @@ class SummarizationAccuracyMetric(Transform):
...
 
         :param record: The input record.
         :returns: The input record with metrics added in.
         """
-        for target_output_key, model_output_key, output_key in zip(
-            self.target_output_keys, self.model_output_keys, self.output_keys
-        ):
-            score = self.compute_metric(record[target_output_key], record[model_output_key])
-            record[output_key] = score
+        if self.target_output_keys is not None:
```
Comment (danielezhu) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:109]: You can reuse the for loop by setting `target_output_list = self.target_output_keys if self.target_output_keys else record[self.target_output_keys_provider]`

File: src/fmeval/transforms/summarization_accuracy_metrics.py [hunk 3]
```diff
@@ -36,67 +37,86 @@ class SummarizationAccuracyMetric(Transform):
 
     def __init__(
         self,
-        target_output_keys: List[str],
-        model_output_keys: List[str],
         output_keys: List[str],
+        model_output_keys: List[str],
         allow_duplicate_input_keys: bool,
+        target_output_keys: Optional[List[str]] = None,
```
Comment (danielezhu) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:43]: Let's keep this as a positional argument so that 1) the argument order doesn't change, which would break things for existing users, and 2) the default configuration isn't invalid; currently, if you use the default options, you'll get a validation error.

For point 1), I really mean that we don't want the signatures of the concrete subclasses to change. Obviously, no one is directly calling this ABC's init method.

File: src/fmeval/transforms/summarization_accuracy_metrics.py [hunk 4]
```diff
@@ -36,67 +37,86 @@ class SummarizationAccuracyMetric(Transform):
...
+        target_output_keys_provider: str = "",
```
Comment (danielezhu) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:44]: We can use `Optional[str]` here to avoid using an empty string as the default value.

File: src/fmeval/transforms/summarization_accuracy_metrics.py [hunk 5]
```diff
@@ -247,23 +277,31 @@ class BertScore(SummarizationAccuracyMetric):
...
         self,
-        target_output_keys: List[str],
-        model_output_keys: List[str],
         output_keys: List[str],
+        model_output_keys: List[str],
         allow_duplicate_input_keys: bool,
-        bertscore_model: Union[BertscoreHelperModel, ActorHandle],
+        target_output_keys: Optional[List[str]] = None,
+        target_output_keys_provider: str = "",
+        bertscore_model: Union[BertscoreHelperModel, ActorHandle] = BertscoreHelperModel(BERTSCORE_DEFAULT_MODEL),
```
Comment (danielezhu) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:285]: We need to put `target_output_keys_provider` as the last argument, or else existing calls to this method will break.

**Review by danielezhu — CHANGES_REQUESTED** [review]

File: src/fmeval/transforms/summarization_accuracy_metrics.py [hunk 6]
```diff
@@ -36,67 +37,82 @@ class SummarizationAccuracyMetric(Transform):
...
+        input_keys = target_output_keys if target_output_keys else [str(target_output_keys_provider)]
         self.register_input_output_keys(
-            target_output_keys + model_output_keys,
+            input_keys + model_output_keys,
             output_keys,
             allow_duplicates=allow_duplicate_input_keys,
         )
         self.target_output_keys = target_output_keys
         self.model_output_keys = model_output_keys
+        self.target_output_keys_provider = str(target_output_keys_provider)
```
Comment (danielezhu) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:97]: Why is this needed?

File: src/fmeval/transforms/summarization_accuracy_metrics.py (hunk 1 above)
Comment (danielezhu) [reply src/fmeval/transforms/summarization_accuracy_metrics.py:148]: It doesn't have a default value. There's no change right now other than adding a type annotation. As an aside, with the new changes, we no longer need this argument. But to keep things backward-compatible, we should leave it for now.

File: src/fmeval/transforms/summarization_accuracy_metrics.py (hunk 5 above)
Comment (danielezhu) [reply src/fmeval/transforms/summarization_accuracy_metrics.py:287]: Discussed with @kirupang-code offline. We need to use this ordering (which is backwards-incompatible) in order for deserialization of these object instances to work properly (since we rely on a specific ordering of the arguments to `__init__`). The existing calls to `BertScore.__init__` in our own code have been updated as necessary, but users who interact with `BertScore` objects directly will be impacted.

We should keep track of this before the next release.

File: test/unit/transforms/test_summarization_accuracy_metrics.py [hunk 7]
```diff
@@ -128,6 +128,31 @@ def test_bert_score_call_with_bertscore_model_object():
     mock_bertscore_model.get_helper_scores.assert_called_once_with("Hello there!", "Hi")
 
 
+def test_bert_score_call_with_target_output_keys_provider():
```
Comment (danielezhu) [inline test/unit/transforms/test_summarization_accuracy_metrics.py:131]: Please add a unit test where there are multiple target outputs, and verify that the max score is returned.

**Review by danielezhu — CHANGES_REQUESTED** [review]

File: src/fmeval/transforms/summarization_accuracy_metrics.py
Comment (kirupang-code) [inline src/fmeval/transforms/summarization_accuracy_metrics.py:1]: Had to add `# type: ignore` to target_output_keys_provider because of its Optional type since we validate that its value is not None.  Errors without the `# type: ignore`:
```python
3. Mypy
=======
src/fmeval/transforms/summarization_accuracy_metrics.py:89:69: error: List item 0 has incompatible type "str | None"; expected "str"  [list-item]
src/fmeval/transforms/summarization_accuracy_metrics.py:111:25: error: Invalid index type "str | None" for "dict[str, Any]"; expected type "str"  [index]
```

**Review by xiaoyi-cheng — DISMISSED** [review]
lgtm

File: test/unit/transforms/test_summarization_accuracy_metrics.py [hunk 8]
```diff
@@ -128,6 +128,49 @@ def test_bert_score_call_with_bertscore_model_object():
...
+        allow_duplicate_input_keys=False,
+        target_output_keys_provider="target_output",
+        bertscore_model=mock_bertscore_model,
+    )
+    sample = {"target_output": ["Hello there!"], "model_output": "Hi"}
+    bs(sample)
+    mock_bertscore_model.get_helper_scores.assert_called_once_with("Hello there!", "Hi")
+
+
+def test_bertscore_multiple_targets_max_score():
```
Comment (danielezhu) [inline test/unit/transforms/test_summarization_accuracy_metrics.py:156]: Please mock the underlying `BertScoreHelperModel` object so that you can have it return known values like 0,1, 0.2, 0.3. Then you can assert that `output["bertscore"]` is exactly equal to the max of your hard-coded values.

**Review by danielezhu — CHANGES_REQUESTED** [review]

**Review by danielezhu — APPROVED** [review]

**Review by xiaoyi-cheng — APPROVED** [review]
